From f02375a88bce5f668ab6aeb2b6192eaf2371c77c Mon Sep 17 00:00:00 2001
From: Jiafei Pan <Jiafei.Pan@nxp.com>
Date: Wed, 25 Sep 2024 17:33:47 +0800
Subject: [PATCH 06/13] plat: imx8m: change dram dvfs to use intercore service

1. Change dvfs to use common intercore service, So that the same SGI irq
   can be shared with SIP cpu_off service.

2. Bring all running Core into WFE
   In general, the requriement for DRAM frequency changing is raised by Linux,
   the current implementation is only bring all the CPU Core used by Linux to
   be WFE state in order to DDR access in the bus frequency changing process.

   But in a flexible AMP system, some Cortex-A Core maybe run RTOS or baremetal,
   these CPU cores should also go into WFE state in the bus frequency changing
   process

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Upstream-Status: Pending
---
 plat/imx/imx8m/ddr/dram.c | 46 ++++++++++++++++++++++-----------------
 1 file changed, 26 insertions(+), 20 deletions(-)

diff --git a/plat/imx/imx8m/ddr/dram.c b/plat/imx/imx8m/ddr/dram.c
index 16507c9f7..7e2f94e18 100644
--- a/plat/imx/imx8m/ddr/dram.c
+++ b/plat/imx/imx8m/ddr/dram.c
@@ -1,17 +1,17 @@
 /*
- * Copyright 2019-2023 NXP
+ * Copyright 2019-2024, 2026 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
-#include <bl31/interrupt_mgmt.h>
 #include <common/runtime_svc.h>
+#include <lib/el3_runtime/cpu_data.h>
 #include <lib/mmio.h>
 #include <lib/spinlock.h>
-#include <plat/common/platform.h>
 
 #include <dram.h>
 #include <gpc.h>
+#include <imx_intercore_svc.h>
 
 #define IMX_SIP_DDR_DVFS_GET_FREQ_COUNT		0x10
 #define IMX_SIP_DDR_DVFS_GET_FREQ_INFO		0x11
@@ -181,18 +181,11 @@ void dram_phy_init(struct dram_timing_info *timing)
 	}
 }
 
-/* EL3 SGI-8 IPI handler for DDR Dynamic frequency scaling */
 static uint64_t waiting_dvfs(uint32_t id, uint32_t flags,
 				void *handle, void *cookie)
 {
 	uint64_t mpidr = read_mpidr_el1();
 	unsigned int cpu_id = MPIDR_AFFLVL0_VAL(mpidr);
-	uint32_t irq;
-
-	irq = plat_ic_acknowledge_interrupt();
-	if (irq < 1022U) {
-		plat_ic_end_of_interrupt(irq);
-	}
 
 	/* set the WFE done status */
 	spin_lock(&dfs_lock);
@@ -215,7 +208,6 @@ void dram_info_init(unsigned long dram_timing_base)
 {
 	uint32_t ddrc_mstr, current_fsp;
 	unsigned int idx = 0;
-	uint32_t flags = 0;
 	uint32_t rc;
 	unsigned int i;
 
@@ -267,8 +259,7 @@ void dram_info_init(unsigned long dram_timing_base)
 	}
 
 	/* Register the EL3 handler for DDR DVFS */
-	set_interrupt_rm_flag(flags, NON_SECURE);
-	rc = register_interrupt_type_handler(INTR_TYPE_EL3, waiting_dvfs, flags);
+	rc = register_intercore_dvfs_svc(waiting_dvfs);
 	if (rc != 0) {
 		panic();
 	}
@@ -332,13 +323,19 @@ int dram_dvfs_get_freq_info(void *handle, u_register_t index)
 	}
 }
 
+static inline aff_info_state_t dvfs_get_aff_info_state_by_idx(unsigned int idx)
+{
+	return get_cpu_data_by_index(idx,
+				     psci_svc_cpu_data.aff_info_state);
+}
+
 int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 	u_register_t x1, u_register_t x2, u_register_t x3)
 {
 	uint64_t mpidr = read_mpidr_el1();
 	unsigned int cpu_id = MPIDR_AFFLVL0_VAL(mpidr);
 	unsigned int fsp_index = x1;
-	uint32_t online_cores = x2;
+	uint32_t running_cores = 0;
 
 	if (x1 == IMX_SIP_DDR_DVFS_GET_FREQ_COUNT) {
 		SMC_RET1(handle, dram_info.num_fsp);
@@ -348,13 +345,23 @@ int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 		wait_ddrc_hwffc_done = true;
 		dsb();
 
-		/* trigger the SGI IPI to info other cores */
-		for (int i = 0; i < PLATFORM_CORE_COUNT; i++) {
-			if (cpu_id != i && (online_cores & (0x1 << (i * 8)))) {
-				plat_ic_raise_el3_sgi(0x8, i);
+		/* Check all running CPU Cores */
+		for (unsigned int i = 0; i < PLATFORM_CORE_COUNT; i++) {
+			/* Skip current CPU Core */
+			if (i == cpu_id) {
+				continue;
+			}
+			/* Perform cache maintanence ahead of reading the target CPU state */
+			flush_cpu_data_by_index(i, psci_svc_cpu_data.aff_info_state);
+			if (dvfs_get_aff_info_state_by_idx(i) == AFF_STATE_ON) {
+				running_cores |= (0x1 << (i * 8));
 			}
 		}
+
+		/* trigger the SGI IPI to info other cores */
+		trigger_intercore_dvfs_svc(running_cores);
 #if defined(PLAT_imx8mq)
+		uint32_t online_cores = x2;
 		for (unsigned int i = 0; i < PLATFORM_CORE_COUNT; i++) {
 			if (i != cpu_id && online_cores & (1 << (i * 8))) {
 				imx_gpc_core_wake(1 << i);
@@ -362,9 +369,8 @@ int dram_dvfs_handler(uint32_t smc_fid, void *handle,
 		}
 #endif
 		/* make sure all the core in WFE */
-		online_cores &= ~(0x1 << (cpu_id * 8));
 		while (1) {
-			if (online_cores == wfe_done) {
+			if (running_cores == wfe_done) {
 				break;
 			}
 		}
-- 
2.43.0

